lib/CL/devices/proxy/pocl_proxy.cpp

```diff
@@ -1164,24 +1197,42 @@ pocl_proxy_build_binary (cl_program program, cl_uint device_i,
   strcat (options, " -cl-kernel-arg-info");
   POCL_MSG_PRINT_PROXY ("BINARY BUILD: options %s \n", options);
 
-  // TODO should binary be already loaded ?
-  assert (program->pocl_binaries[device_i]);
-  char program_bc_path[POCL_MAX_PATHNAME_LENGTH];
-  pocl_cache_program_bc_path (program_bc_path, program, device_i);
-
-  assert (pocl_exists (program_bc_path));
-  assert (program->binaries[device_i]);
-  assert (program->binary_sizes[device_i] > 0);
-  const unsigned char *binary = program->binaries[device_i];
-  size_t size = program->binary_sizes[device_i];
-
   assert (link_program);
-
-  // context, num_devices, devices, sizes, binaries, binary_statuses, &err);
   cl_int status = CL_INVALID_VALUE;
-  size_t temp_size = (size_t)size;
-  prog = clCreateProgramWithBinary (context, 1, &d->device_id, &temp_size,
-                                    &binary, &status, &err);
+
+  if (spir_build) {
+    assert(program->program_il != NULL);
+    assert(program->program_il_size > 0);
+    prog = clCreateProgramWithIL(context, (const void *)program->program_il,
+                                 program->program_il_size, &err);
+
+    /* put spir-v in cache if successful */
+    if (err == CL_SUCCESS) {
+      char spirv_path[POCL_MAX_PATHNAME_LENGTH];
+      pocl_cache_tempname(spirv_path, ".spv", NULL);
+      err = pocl_write_file(spirv_path, program->program_il,
+                            program->program_il_size, 0);
+      POCL_RETURN_ERROR_ON((err != 0), CL_BUILD_PROGRAM_FAILURE,
+                           "failed to write the SPIR-V file into cache\n");
+      pocl_cache_create_program_cachedir(program, device_i, program->program_il,
+                                         program->program_il_size, spirv_path);
+    }
+  } else {
+    // TODO should binary be already loaded ?
+    assert(program->pocl_binaries[device_i]);
+    char program_bc_path[POCL_MAX_PATHNAME_LENGTH];
+    pocl_cache_program_bc_path(program_bc_path, program, device_i);
+
+    assert(pocl_exists(program_bc_path));
```

Comment:
Preferably, do not use asserts for error checking. Return an error (with an error message) and let the caller deal with it.

Reply:
This specific piece of code already existed and just got moved around, but l now that i'm digging deeper into it, I'm not even sure if it makes sense, so i'll have to test this.